Simple item smashing with an anvil

tommyTT opened the pull request:
I re-implemented the system to smash items with the anvil (mostly) according to the specs we discussed on discord:
- Multiple inputs (specific item or ore dict name).
- Multiple outputs.
- Requirement for minimum fall height of the anvil.
- Requirement for specific state of the block the anvil lands on.
- Separate JEI category for the recipes with a very basic textual representation of height and block requirements.

I implemented everything similar to the cauldron recipes and used RecipeMatch. However I needed to implement a custom sub class in order to allow multiple inputs. Maybe this is something that already exists and I didn't find it? Is there a better way to implement it? Otherwise that class could also be moved to Mantle.

The JEI implementation is very basic at the moment. Because of limited space it is limited to only display 3 inputs and 3 outputs. The requirements for the fall height and the block state are shown via tooltips.

I also included a few example recipes in the default config to demonstrate and test the configuration. 

Please let me know what should be changed, especially the logic for applying the recipes I'm not too sure about. I included a hard limit for the number of times recipes are applied per "anvil operation" in order to limit the number of item entities as well as the computation time.

This PR replaces #69 which can be closed.

tommyTT committed:
re-implemented a simple version of the item smashing

tommyTT committed:
added simple JEI display for the recipes

tommyTT committed:
- added ability to specify ore dict input
- added ability to specify multiple output stacks
- cleaned up a few methods that are not used
- implemented the JEI recipes as a seperate category
- i18n

tommyTT committed:
implemented ability to specify multiple inputs
move JEI recipes for item smashing to its own wrapper and include block state requirements
added a few more example recipes and comments for them

tommyTT committed:
limit the amount of recipes to apply in one operation

KnightMiner commented on src/main/java/knightminer/inspirations/common/Config.java:
```diff
@@ -592,6 +618,166 @@ private static void processAnvilSmashing(String[] transformations) {
+	 * @param input input string
+	 * @return the block state or null if none could be found
+	 */
+	private static IBlockState parseBlockstate(String input) {
```
Should be in Utils

KnightMiner commented on src/main/java/knightminer/inspirations/common/Config.java:
```diff
@@ -592,6 +618,166 @@ private static void processAnvilSmashing(String[] transformations) {
+	 * @param input input string
+	 * @return item stack, might be {@link ItemStack#EMPTY} if it cannot be parsed
+	 */
+	private static ItemStack parseItemStackWithCount(String input) {
```
Util or RecipeUtil

KnightMiner commented on src/main/java/knightminer/inspirations/common/Config.java:
```diff
@@ -592,6 +618,166 @@ private static void processAnvilSmashing(String[] transformations) {
+	 * @param value string
+	 * @return the integer or null if it cannot be parsed
+	 */
+	private static Integer parseInteger(String value) {
```
Probably Util

KnightMiner commented on src/main/java/knightminer/inspirations/library/recipe/anvil/IAnvilRecipe.java:
```diff
@@ -0,0 +1,35 @@
+	 * @param state  State of the block the anvil landed on
+	 * @return
+	 */
+	NonNullList<ItemStack> transformInput(NonNullList<ItemStack> stack, int height, IBlockState state);
```
I would call this getOutput(s). transformInput was just used with cauldrons as it processed both reducing the input stack size and the normal output sepearate

KnightMiner commented on src/main/java/knightminer/inspirations/library/recipe/anvil/ISimpleAnvilRecipe.java:
```diff
@@ -0,0 +1,43 @@
+	 *
+	 * @return Input block state
+	 */
+	default Object getInputState() {
```
Output can just be a blockstate, cauldron used Object because I parsed the cauldron state into another class in the simple cauldron recipe. Since there is no output state `getState()` is probably fine

KnightMiner commented on src/main/java/knightminer/inspirations/library/recipe/anvil/ISimpleAnvilRecipe.java:
```diff
@@ -0,0 +1,43 @@
+	 *
+	 * @return
+	 */
+	default Integer getFallHeight() {
```
This uses class integer, but does not explain what null means. Should say something like "Null means fall height is ignored"

KnightMiner commented on src/main/java/knightminer/inspirations/recipes/block/BlockSmashingAnvil.java:
```diff
@@ -21,11 +34,115 @@ public BlockSmashingAnvil() {
+		List<ItemStack> results = new ArrayList<>();
+		boolean recipeApplied = false;
+		int iterations = 0;
+		while(iterations++ < MAX_RECIPE_APPLICATIONS) {
```
I would rather not be scanning the recipe registry many times in a single landing. Ideally just one recipe would catch then we reuse it until we have our results, so if we have too many inputs for a recipe it just does nothing.

It might also be a good idea to have the recipe just handle stacks using math instead of a loop. So for somethimg like the bone example, instead of saying the result is bonemeal * 4, its bonemeal * 4 * inputStackSize.

KnightMiner commented on src/main/java/knightminer/inspirations/plugins/jei/smashing/SmashingItemRecipeWrapper.java:
```diff
@@ -0,0 +1,74 @@
+
+		Integer minFallHeight = recipe.getFallHeight();
+		this.heightRequirementString =
+				minFallHeight != null ? Util.translateFormatted("gui.jei.anvil_smashing.height", minFallHeight) : null;
```
I'm really not a fan of this being just a tooltip. I would rather draw a string of the fall height

KnightMiner commented on src/main/java/knightminer/inspirations/plugins/jei/smashing/SmashingItemRecipeWrapper.java:
```diff
@@ -0,0 +1,74 @@
+				minFallHeight != null ? Util.translateFormatted("gui.jei.anvil_smashing.height", minFallHeight) : null;
+
+		Object inputState = recipe.getInputState();
+		this.blockRequirementString = inputState instanceof IBlockState ?
```
Might be nice to have this show visually in JEI, maybe place this under the item that is under the anvil? Or alternatively, what about if we place the items on top of the anvil and the blocks underneath, to make it extra clear which are items

KnightMiner commented:
I looked at this briefly and left a few comments on improvements. Overall I like this approach a lot better.

----

One thing I am considering is if it might actually be a good idea to combine both anvil recipes into one, so in addition to the item output, have a state output (which could be the same as the input, maybe use null to signify no change). This means the block smashing functions could be redirected to the item smashing object now (which could be just a smashing wrapper), and the possibility of smashing an item into a block to change it exists (like we smash blaze powder into stone and it becomes heated stone)

I am not sure how I feel about multiple recipes running when landing. I like a stack of items (like a stack of bones) all processing if any are there, but if you have multiple different inputs its too easy to get a race condition (if one recipe takes ender pearls and one takes ender pearls and blaze powder). Maybe have the recipe only match if the items list matches exactly? This would mean too many items would prevent a recipe. If combined with above, block smashing would be dependent upon there being no items (though we could after running the item smash check try a second time with an empty list of items to get block smashing that ignores items)

tommyTT committed:
refactoring and cleanup
implemented the JEI visualization a bit differently

tommyTT committed:
allowed multiples for ore dict inputs
fixed JEI representation of ore dict entries
changed logic of how to apply the recipes
